onnxruntime/core/graph/graph.cc

```diff
@@ -2055,6 +2056,39 @@ void Graph::RemoveInitializedTensor(const std::string& tensor_name) {
   }
 }
 
+Status Graph::ReplaceInitializedTensor(const ONNX_NAMESPACE::TensorProto& new_initializer) {
+  const auto& initializer_name = new_initializer.name();
+  auto name_to_initializer_it = name_to_initial_tensor_.find(initializer_name);
+  ORT_RETURN_IF_NOT(name_to_initializer_it != name_to_initial_tensor_.end(),
+                    "Failed to find existing initializer with name ", initializer_name, ".");
+
+  assert(name_to_initializer_it->second);
+  const auto& old_initializer = *(name_to_initializer_it->second);
+
+  auto dims_eq = [&old_initializer, &new_initializer]() {
+    if (old_initializer.dims_size() != new_initializer.dims_size()) return false;
+    for (int i = 0; i < old_initializer.dims_size(); ++i) {
+      if (old_initializer.dims(i) != new_initializer.dims(i)) return false;
+    }
+    return true;
+  };
+
+  ORT_RETURN_IF_NOT(dims_eq(), "Replacement tensor's dimensions do not match.");
+  ORT_RETURN_IF_NOT(old_initializer.data_type() == new_initializer.data_type(),
+                    "Replacement tensor's data type does not match.");
+
+  auto& mutable_initializers = *graph_proto_->mutable_initializer();
+  auto old_mutable_initializer_ptr_it = std::find(
+      mutable_initializers.pointer_begin(), mutable_initializers.pointer_end(), &old_initializer);
+  assert(old_mutable_initializer_ptr_it != mutable_initializers.pointer_end());
```

Comment:
>assert [](start = 2, length = 6)

assert is no-op for release build. Use ORT_ENFORCE instead.

Reply:
Yes, this is for catching programming errors so I think assert is appropriate. I'd prefer to consistently return status values or throw exceptions instead of mixing them.

Reply:
Maybe use ORT_ENFORCE_DEBUG()? [https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/codegen/common/common.h#L10-L14](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/codegen/common/common.h#L10-L14)

IMO It's better to have consistent error/exception handling, even for debug code.

---
In reply to: [337784729](https://github.com/microsoft/onnxruntime/pull/2230#discussion_r337784729) [](ancestors = 337784729)

Reply:
Thanks for the suggestions.

In this case, I'd still prefer something that fails more drastically than an exception (e.g., ORT_ENFORCE[_DEBUG]). For one thing, exceptions can be caught and ignored. I think it helps to distinguish between different types of error conditions, including but not limited to errors the user can do something about (bad input/value not present/...) and implementation errors (bugs). Providing an actionable indication of failure (via exception or return value) seems appropriate for the former, while something which fails fast (like assert or terminate) seems better for the latter.

Reply:
For consistency with other things here, ORT_RETURN_IF_NOT would be better. There may be an unexpected sequence of events that leads to things getting out of sync, which unit tests may not catch given you're dealing with two separate pieces (the GraphProto and the name_to_initial_tensor_ map).

Also consider if it should be an error. If we consider name_to_initial_tensor_ to be the most correct at runtime, if the TensorProto isn't in the GraphProto why not just add it?

Reply:
Ok, changed to ORT_ENFORCE. I'll try to get a better understanding of when this should be used.